src/System.Linq/src/System/Linq/Select.cs

```diff
@@ -226,6 +258,17 @@ public List<TResult> ToList()
 
             public int GetCount(bool onlyIfCheap)
             {
+                // In case someone uses Count() to force evaluation of
+                // the selector, run it provided `onlyIfCheap` is false.
+
+                if (!onlyIfCheap)
+                {
+                    foreach (TSource item in _source)
+                    {
+                        _selector(item);
+                    }
+                }
```

Comment:
IIRC, if we just returned `-1` in this case then the calling `Count()` method would do pretty much the above. I would imagine this would be slightly faster, but only slightly (that is just a guess though). Do we need the extra code here?